tpm2/tpm2.go

```diff
@@ -0,0 +1,1247 @@
+	cmd, err := encodeStartAuthSession(tpmKey, bindKey, nonceCaller, secret, se, sym, hashAlg)
+	if err != nil {
+		return 0, nil, err
+	}
+	resp, err := runCommand(rw, tagNoSessions, cmdStartAuthSession, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return 0, nil, err
+	}
+	handle, nonce, err := decodeStartAuthSession(resp)
+	if err != nil {
+		return 0, nil, fmt.Errorf("decoding StartAuthSession response: %v", err)
+	}
+	return handle, nonce, nil
+}
+
+func encodeUnseal(itemHandle Handle, password string, sessionHandle Handle) ([]byte, error) {
+	b1, err := tpmutil.Pack(itemHandle, []byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b2, err := encodePasswordAuthArea(password, sessionHandle)
+	if err != nil {
+		return nil, err
+	}
+	return append(b1, b2...), nil
+}
+
+func decodeUnseal(in []byte) ([]byte, []byte, error) {
+	var unsealed []byte
+	var digest []byte
+
+	err := tpmutil.Unpack(in[4:], &unsealed, &digest)
+	if err != nil {
+		return nil, nil, err
+	}
+	return unsealed, digest, nil
+}
+
+// Unseal returns the data for a loaded sealed object.
+// Returns unsealed data and a digest.
+func Unseal(rw io.ReadWriter, itemHandle Handle, password string, sessionHandle Handle, digest []byte) ([]byte, []byte, error) {
+	cmd, err := encodeUnseal(itemHandle, password, sessionHandle)
+	if err != nil {
+		return nil, nil, err
+	}
+	resp, err := runCommand(rw, tagSessions, cmdUnseal, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return nil, nil, err
+	}
+	unsealed, nonce, err := decodeUnseal(resp)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decoding Unseal response: %v", err)
+	}
+	return unsealed, nonce, nil
+}
+
+func encodeQuote(signingHandle Handle, parentPassword, ownerPassword string, toQuote []byte, pcrNums []int, sigAlg Algorithm) ([]byte, error) {
+	b1, err := tpmutil.Pack(signingHandle)
+	if err != nil {
+		return nil, err
+	}
+	b2, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b3, err := encodePasswordAuthArea(parentPassword, HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	b4, err := tpmutil.Pack(toQuote, sigAlg)
+	if err != nil {
+		return nil, err
+	}
+	b5, err := encodeLongPCR(1, pcrNums)
+	if err != nil {
+		return nil, err
+	}
+	args := append(b1, b2...)
+	args = append(args, b3...)
+	args = append(args, b4...)
+	args = append(args, b5...)
+	return args, nil
+}
+
+func decodeQuote(in []byte) ([]byte, uint16, uint16, []byte, error) {
+	var empty []byte
+	var buf []byte
+	var attest []byte
+	var signature []byte
+	var s1 uint16
+	var s2 uint16
+
+	err := tpmutil.Unpack(in, &empty, &buf)
+	if err != nil {
+		return nil, 0, 0, nil, err
+	}
+
+	err = tpmutil.Unpack(buf, &attest, &s1, &s2, &signature)
+	if err != nil {
+		return nil, 0, 0, nil, err
+	}
+	return attest, s1, s2, signature, nil
+}
+
+// Quote returns a quote of PCR values.
```

Comment:
Maybe mention here what a Quote is. Even a reference to the TPM docs would be useful for readers.

Reply:
Added a small clarification, but this package is full of TPM lingo (PCR, Handle, Session etc).
I'm not sure how much detail should the package provide, it should be assumed that reader has some understanding of TPMs.

That being said, there is no good concise learning material for TPM 2.0, so perhaps more detail is good.